### ISSUE TITLE
FlexGen modification of script for low memory systems and maybe add tcmalloc to the whole sytem?

### ISSUE THREAD

**Issue by angrysky56** [oobabooga/text-generation-webui#2480]
I am not anything close to a coder but I was trying to get a model converted that was opt 13b and the process was killed after a while so I wanted to try to make the code take longer if necessary but get through- I have not even tried this since it comes from Bing but it is the idea to allow batch size selection and add virtual ram if needed.
This is for Linux and probably wont work- the AI made some suggestions too.

1 Reduce the batch size: You have already added a batch size for processing model weights. You can try reducing the batch size further to reduce memory usage.

2 Use a memory profiler: You can use a memory profiler such as tracemalloc or memory_profiler to identify parts of your code that use a large amount of memory. This can help you optimize those parts of your code to reduce memory usage.

a solution I got from automatic1111 that sped up model loading from minutes to seconds for that ui-
Yes, the commands you provided can be used to compile a C program with tcmalloc, which is a memory allocator that can improve the performance of memory allocation in your program. However, these commands are specific to C programs and cannot be directly applied to your Python script.

If you want to use tcmalloc with your Python script, you can try setting the LD_PRELOAD environment variable to the path of the libtcmalloc_minimal.so library before running your script. This will cause the dynamic linker to load tcmalloc and use it for memory allocation in your script. Here’s an example of how you can do this:

export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so.4
python3 your_script.py

in my case 

export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so.4.5.9
python3 convert-to-flexgenv2batches.py

you can set the LD_PRELOAD environment variable before running your Python script to use tcmalloc for memory allocation. Here’s an example of how you can do this in a bash shell:

export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so.4.5.9
python3 convert-to-flexgenv2batches.py

In this example, the first line sets the LD_PRELOAD environment variable to the path of the libtcmalloc_minimal.so library. The second line runs your Python script with tcmalloc enabled for memory allocation.

Keep in mind that the exact path to the libtcmalloc_minimal.so library may vary depending on your system.

3 Free up memory: You can try freeing up memory in your script by deleting large objects that are no longer needed and calling gc.collect() to run the garbage collector. (I didn't try to add this)

4 Use a different data type: If your model uses 32-bit floating-point numbers (float32), you can try converting it to use 16-bit floating-point numbers (float16) to reduce memory usage. However, this may affect the accuracy of the model.

Attempted to adapt from here-
https://github.com/oobabooga/text-generation-webui/blob/main/docs/FlexGen.md

import argparse
import os
from pathlib import Path

import numpy as np
import torch
from tqdm import tqdm
from transformers import AutoModelForCausalLM, AutoTokenizer

import subprocess

def create_swapfile():
    subprocess.run(["sudo", "fallocate", "-l", "8G", "/swapfile"])
    subprocess.run(["sudo", "chmod", "600", "/swapfile"])
    subprocess.run(["sudo", "mkswap", "/swapfile"])
    subprocess.run(["sudo", "swapon", "/swapfile"])

parser = argparse.ArgumentParser(formatter_class=lambda prog: argparse.HelpFormatter(prog, max_help_position=54))
parser.add_argument('MODEL', type=str, default=None, nargs='?', help="Path to the input model.")
args = parser.parse_args()

def disable_torch_init():
    """
    Disable the redundant torch default initialization to accelerate model creation.
    """
    import torch
    global torch_linear_init_backup
    global torch_layer_norm_init_backup

    torch_linear_init_backup = torch.nn.Linear.reset_parameters
    setattr(torch.nn.Linear, "reset_parameters", lambda self: None)

    torch_layer_norm_init_backup = torch.nn.LayerNorm.reset_parameters
    setattr(torch.nn.LayerNorm, "reset_parameters", lambda self: None)

def restore_torch_init():
    """Rollback the change made by disable_torch_init."""
    import torch
    setattr(torch.nn.Linear, "reset_parameters", torch_linear_init_backup)
    setattr(torch.nn.LayerNorm, "reset_parameters", torch_layer_norm_init_backup)

if __name__ == '__main__':
    path = Path(args.MODEL)
    model_name = path.name

    print(f"Loading {model_name}...")
    # disable_torch_init()
    model = AutoModelForCausalLM.from_pretrained(path, torch_dtype=torch.float16, low_cpu_mem_usage=True)
    # restore_torch_init()

    tokenizer = AutoTokenizer.from_pretrained(path)

    out_folder = Path(f"models/{model_name}-np")
    if not Path(out_folder).exists():
        os.mkdir(out_folder)

    print(f"Saving the converted model to {out_folder}...")
    
    # Set batch size for processing model weights
    batch_size = 10
    
    for name, param in tqdm(list(model.model.named_parameters())):
        name = name.replace("decoder.final_layer_norm", "decoder.layer_norm")
        param_path = os.path.join(out_folder, name)
        
        # Process model weights in batches
        param_data = param.cpu().detach().numpy()
        for i in range(0, len(param_data), batch_size):
            batch = param_data[i:i+batch_size]
            with open(param_path, "ab") as f:
                np.save(f, batch)

**Comment by angrysky56** [comment]
tried it, did not seem to work-

export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so.4.5.9
python3 convert-to-flexgenv2batches.py /home/angrysky/Projects/oobabooga_linux/text-generation-webui/models/KoboldAI_OPT-13B-Erebus/

**Comment by angrysky56** [comment]
tried it with opt 6b with batch size 1 and it took 5 minutes- I don't know if it made a difference, I would have to try the regular script on it.

(/home/angrysky/Projects/oobabooga_linux/installer_files/env) angrysky@angrysky-ROG-STRIX-G10DK-G10DK:~/Projects/oobabooga_linux/text-generation-webui$ export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so.4.5.9
(/home/angrysky/Projects/oobabooga_linux/installer_files/env) angrysky@angrysky-ROG-STRIX-G10DK-G10DK:~/Projects/oobabooga_linux/text-generation-webui$ python3 convert-to-flexgenv2batches.py /home/angrysky/Projects/oobabooga_linux/text-generation-webui/models/facebook_opt-6.7b/
Loading facebook_opt-6.7b...
Loading checkpoint shards: 100%|█████████████████| 2/2 [03:46<00:00, 113.32s/it]
Saving the converted model to models/facebook_opt-6.7b-np...
100%|█████████████████████████████████████████| 516/516 [04:57<00:00,  1.73it/s]
(/home/angrysky/Projects/oobabooga_linux/installer_files/env) angrysky@angrysky-ROG-STRIX-G10DK-G10DK:~/Projects/oobabooga_linux/text-generation-webui$

**Comment by angrysky56** [comment]
Ok I did something wrong- I got the output file but I have no idea what to do to get it to work or compile or...? I loaded it as per instructions but it wasn't reading it as a valid file.
